File: agent/internal/server/protocol.go
package server

import (
	"regexp"
	"strconv"
)

// ProtocolVersion is what the agent and the shell must agree on: the bridge
// methods and the provisioning payload. Bump it only when one of those
// changes incompatibly. See docs/architecture.md, "Protocol version".
const ProtocolVersion = 1

// protocolSource is the source name of the hello frame. It never enters the
// store, so it is not in /health or in a snapshot.
const protocolSource = "protocol"

// shellTokenPattern matches the token the shell appends to the WebView user
// agent, for example "spotdash-shell/0.1.0 proto/1".
var shellTokenPattern = regexp.MustCompile(`(?:^|\s)spotdash-shell/(\S+) proto/(\d+)(?:\s|$)`)

// shellToken reads the shell's version and protocol from a User-Agent. ok is
// false for a browser or any agent string without a valid token.
func shellToken(userAgent string) (version string, protocol int, ok bool) {
	match := shellTokenPattern.FindStringSubmatch(userAgent)
	if match == nil {
		return "", 0, false
	}
	n, err := strconv.Atoi(match[2])
	if err != nil {
		return "", 0, false
	}
	return match[1], n, true
}

File: shell/app/src/main/java/dev/spotdash/shell/Protocol.kt
package dev.spotdash.shell

/**
 * The protocol version handshake with the agent. Logic that needs no Android, kept out of the
 * activity for unit tests. See docs/architecture.md, "Protocol version".
 */

/** What the shell speaks: the bridge methods and the provisioning payload. Bump it only when either changes incompatibly. */
internal const val PROTOCOL_VERSION = 1

/** Which side is behind when the shell and the agent disagree. */
internal enum class ProtocolMismatch { NONE, UPDATE_SHELL, UPDATE_AGENT }

/** The WebView user agent with the shell token appended. The agent and the panel read it back. */
internal fun shellUserAgent(defaultAgent: String, versionName: String): String {
    val version = versionName.replace(Regex("\\s+"), "_")
    return "$defaultAgent spotdash-shell/$version proto/$PROTOCOL_VERSION"
}

private val PROTOCOL_FIELD = Regex("\"protocol\"\\s*:\\s*(\\d+)(?![\\d.])")

/** The `protocol` from a `/health` body, or null when it is absent or not a whole number. */
internal fun agentProtocol(healthBody: String): Int? =
    PROTOCOL_FIELD.find(healthBody)?.groupValues?.get(1)?.toIntOrNull()

/** An agent that reports nothing is an older build and is not called a mismatch. */
internal fun protocolMismatch(shell: Int, agent: Int?): ProtocolMismatch = when {
    agent == null || agent == shell -> ProtocolMismatch.NONE
    shell < agent -> ProtocolMismatch.UPDATE_SHELL
    else -> ProtocolMismatch.UPDATE_AGENT
}
